core/src/main/java/com/github/philippheuer/events4j/core/EventManager.java

```diff
@@ -177,14 +188,34 @@ public boolean hasEventHandler(Class eventHandlerClass) {
     }
 
     /**
-     * Registers a new consumer based default event handler if supported
+     * Registers a event consumer
      *
      * @param eventClass the event class to obtain events from
      * @param consumer   the event consumer / handler method
      * @param <E>        the event type
      * @return a new Disposable of the given eventType
      */
-    public <E extends Object> IDisposable onEvent(Class<E> eventClass, Consumer<E> consumer) {
+    public <E> IEventSubscription onEvent(Class<E> eventClass, Consumer<E> consumer) {
+        return onEvent(consumer.getClass().getCanonicalName()+"/"+consumerSequence.getAndAdd(1), eventClass, consumer);
+    }
+
+    /**
+     * Registers a event consumer and assigns a id
+     * <p>
+     * This method will return null if the id was already used.
+     *
+     * @param id         unique id used to identify this consumer
+     * @param eventClass the event class to obtain events from
+     * @param consumer   the event consumer / handler method
+     * @param <E>        the event type
+     * @return           a new Disposable of the given eventType
+     */
+    public <E> IEventSubscription onEvent(String id, Class<E> eventClass, Consumer<E> consumer) {
+        // return null if a disposable with the given id is already present when idUnique is set
+        if (activeSubscriptions.containsKey(id)) {
```

Comment:
Since the subscription isn't added immediately and synchronization hasn't been applied, this could be problematic if onEvent is called concurrently; imagine two threads registering a subscription with the same id for the first time. It could be possible for both threads to see containsKey returning false before they each go on to update the map

Reply:
Good point, i made `.onEvent` `synchronized` because it's only running a few ns to register a listener. They are added to the activeSubscriptions inside the constructors of the disposable wrappers.